File: android/folkbot/src/main/java/kallaspriit/bot/WebsocketTransport.java
package kallaspriit.bot;

import android.util.Log;

import org.java_websocket.WebSocket;
import org.java_websocket.drafts.Draft;
import org.java_websocket.handshake.ClientHandshake;

import java.net.InetSocketAddress;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class WebsocketTransport extends org.java_websocket.server.WebSocketServer implements Transport {
  private static final String TAG = "WebsocketTransport";
  private List<Transport.Listener> listeners = new ArrayList<>();
  private Transport.State state = State.DISCONNECTED;

  WebsocketTransport(int port, Draft draft) {
    super(new InetSocketAddress(port), Collections.singletonList(draft));

    this.setReuseAddr(true);

    // disable batching algorithm to reduce latency
    this.setTcpNoDelay(true);

    setState(State.CONNECTING);
  }

  @Override
  public String getName() {
    return "WebSocket";
  }

  public void addListener(Transport.Listener listener) {
    this.listeners.add(listener);
  }

  @Override
  public void onOpen(WebSocket connection, ClientHandshake handshake) {
    Log.i(TAG, "client connected");

    listeners.forEach(listener -> listener.onClientConnected(this));
  }

  @Override
  public void onClose(WebSocket connection, int code, String reason, boolean remote) {
    Log.i(TAG, "client disconnected");
  }

  @Override
  public void onError(WebSocket connection, Exception error) {
    Log.w(TAG, "error: " + error.getMessage());

    error.printStackTrace();

    listeners.forEach(listener -> listener.onError(this, error));
  }

  @Override
  public void onStart() {
    Log.i(TAG, "web-socket server started");

    setState(Transport.State.CONNECTED);
  }

  @Override
  public void onMessage(WebSocket connection, String message) {
    Log.d(TAG, "got message: '" + message + "'");

    listeners.forEach(listener -> listener.onMessageReceived(this, message));
  }

  @Override
  public void send(String message) {
    this.broadcast(message);

    listeners.forEach(listener -> listener.onMessageSent(this, message));
  }

  private void setState(State newState) {
    if (newState == state) {
      return;
    }

    State previousState = state;

    state = newState;

    listeners.forEach(listener -> listener.onStateChanged(this, newState, previousState));
  }
}

File: app/src/lib/transport/WebsocketTransport.ts
import { dummyLogger, Logger } from "ts-log";

import { Transport, TransportListener, TransportStatus } from "./Transport";

export interface WebsocketTransportOptions {
  host: string;
  port: number;
  useSSL?: boolean;
  reconnectInterval?: number;
  log?: Logger;
}

export class WebsocketTransport implements Transport {
  private readonly listeners: TransportListener[] = [];
  private options: Required<WebsocketTransportOptions>;
  private state: TransportStatus = TransportStatus.DISCONNECTED;
  private wasConnected = false;
  private ws?: WebSocket;
  private reconnectTimeout?: NodeJS.Timeout;

  constructor(options: WebsocketTransportOptions) {
    this.options = {
      useSSL: false,
      log: dummyLogger,
      reconnectInterval: 1000,
      ...options,
    };
  }

  get log() {
    return this.options.log;
  }

  getName() {
    return "WebSocket";
  }

  isAvailable() {
    return WebSocket !== undefined;
  }

  getOptions() {
    return this.options;
  }

  updateOptions(options: Partial<WebsocketTransportOptions>) {
    this.options = {
      ...this.options,
      ...options,
    };
  }

  getState() {
    return this.state;
  }

  addListener(listener: TransportListener) {
    this.listeners.push(listener);
  }

  removeListener(listener: TransportListener) {
    const listenerIndex = this.listeners.findIndex((existingListener) => listener === existingListener);

    if (listenerIndex === -1) {
      return false;
    }

    this.listeners.splice(listenerIndex, 1);

    return true;
  }

  async connect() {
    const url = `${this.options.useSSL ? "wss" : "ws"}://${this.options.host}:${this.options.port}`;

    this.log.info(`connecting to web-socket server at ${url}`);

    // close existing connection if exists
    if (this.ws) {
      console.log("closing connection", this.ws);

      // clear event handlers
      this.ws.onopen = () => {};
      this.ws.onclose = () => {};
      this.ws.onerror = () => {};
      this.ws.onmessage = () => {};

      // close the connection
      this.ws.close();

      this.ws = undefined;
    }

    // clear reconnect timeout if exists
    if (this.reconnectTimeout !== undefined) {
      clearTimeout(this.reconnectTimeout);

      this.reconnectTimeout = undefined;
    }

    // update state depending on whether the connection was ever established
    this.setState(this.wasConnected ? TransportStatus.RECONNECTING : TransportStatus.CONNECTING);

    // attempt to open web-socket connection
    this.ws = new WebSocket(url);

    // handle open event
    this.ws.onopen = (_event) => {
      this.log.info("established web-socket connection");

      this.wasConnected = true;

      // update state
      this.setState(TransportStatus.CONNECTED);
    };

    // handle close event
    this.ws.onclose = (event) => {
      const logDetails = `code: ${event.code}, reason: ${event.reason}, was clean: ${event.wasClean ? "yes" : "no"}`;

      if (this.wasConnected) {
        this.log.warn(`connection to web-socket was lost (${logDetails})`);
      } else {
        this.log.warn(`connecting to web-socket failed (${logDetails})`);
      }

      // update state
      this.setState(TransportStatus.DISCONNECTED);

      // only attempt to reconnect if connection has succeeded before
      if (this.wasConnected) {
        this.reconnectTimeout = setTimeout(() => {
          this.reconnectTimeout = undefined;

          void this.connect();
        }, this.options.reconnectInterval);
      }
    };

    // handle error event
    this.ws.onerror = (_event) => {
      this.log.warn("got web-socket error");

      // notify the listeners
      this.listeners.forEach((listener) => listener.onError(this));
    };

    // handle message event
    this.ws.onmessage = (event) => {
      const message = event.data;

      // this.log.info("got message", message);

      // notify the listeners of message received
      this.listeners.forEach((listener) => listener.onMessageReceived(this, message));
    };
  }

  send(message: string) {
    // we can only send messages if we're connected
    if (!this.ws || this.state !== TransportStatus.CONNECTED) {
      this.log.warn(`sending message "${message}" requested but websocket state is ${this.state}`);

      // notify of failed message sending attempt
      this.listeners.forEach((listener) => listener.onMessageSent(this, message, false));

      return false;
    }

    // send the message
    this.ws.send(message);

    // notify the listeners
    this.listeners.forEach((listener) => listener.onMessageSent(this, message, true));

    return true;
  }

  private setState(newState: TransportStatus) {
    // ignore if state did not change
    if (newState === this.state) {
      return;
    }

    // remember previous state
    const previousState = this.state;

    // update current state
    this.state = newState;

    // notify the listeners of state change
    this.listeners.forEach((listener) => listener.onStatusChanged(this, newState, previousState));
  }
}
